src/components/tl/mlx5/mcast/tl_mlx5_mcast_rcache.c

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_mlx5_mcast_rcache.h"
+
+static ucc_status_t ucc_tl_mlx5_mcast_coll_reg_mr(ucc_tl_mlx5_mcast_coll_context_t *ctx,
+                                                  void *data, size_t data_size, void **mr)
+{
+    *mr = ibv_reg_mr(ctx->pd, data, data_size, IBV_ACCESS_LOCAL_WRITE,
+                     IBV_ACCESS_REMOTE_READ, IBV_ACCESS_REMOTE_WRITE);
+
+    tl_debug(ctx->lib, "External memory register: ptr %p, len %zd, mr %p",
+            data,  data_size, (*mr));
+    if (!*mr) {
+        tl_error(ctx->lib, "TL/MCAST: Failed to register MR\n");
+        return UCC_ERR_NO_MEMORY;
+    }
+    
+    return UCC_OK;
+}
+
+
+static ucc_status_t ucc_tl_mlx5_mcast_coll_dereg_mr(ucc_tl_mlx5_mcast_coll_context_t *ctx, void *mr)
+{
+    if(mr != NULL) {
+        tl_debug(ctx->lib, "External memory deregister: mr %p", mr);
+        ibv_dereg_mr(mr);
+    } else {
+        tl_debug(ctx->lib, "External memory mr %p already deregistered", mr);
+    }
+
+    return UCC_OK;
+}
+
+static ucs_status_t
+ucc_tl_mlx5_mcast_rcache_mem_reg_cb(void *context, ucs_rcache_t *rcache,
```

Comment:
we should use `ucc_rcache_t` instead of `ucs_rcache_t`

Reply:
Please note
#define ucc_rcache_ops_t             ucs_rcache_ops_t
These are defined in ucx and call back must return ucs_status_t. Therefore, ucc_tl_mlx5_mcast_rcache_mem_reg_cb must return ucs_status_t otherwise it won't compile